llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -188,54 +189,141 @@ int PostPipeliner::fit(MachineInstr *MI, int First, int Last, int II) {
+  const int Latest = -1;
   for (int K = 0; K < NInstr; K++) {
+    auto &Me = Info[K];
+    Me.Latest = Latest;
     SUnit &SU = DAG->SUnits[K];
-    const int Earliest = Info[K].Earliest;
-    scheduleNode(SU, Earliest);
+    for (auto &Dep : SU.Preds) {
+      if (Dep.getKind() != SDep::Data) {
+        continue;
+      }
+      int P = Dep.getSUnit()->NodeNum;
+      assert(P < K);
+      Me.Ancestors.insert(P);
+      auto &Pred = Info[P];
+      for (int Anc : Pred.Ancestors) {
+        Me.Ancestors.insert(Anc);
+      }
+    }
+    for (auto &Dep : SU.Succs) {
+      auto *Succ = Dep.getSUnit();
+      if (Succ->isBoundaryNode()) {
+        continue;
+      }
+      auto &SInfo = Info[Succ->NodeNum];
+      const int NewEarliest = Me.Earliest + Dep.getSignedLatency();
+      SInfo.Earliest = std::max(SInfo.Earliest, NewEarliest);
+    }
   }
+}
 
-  // Propagate Earliest upstream, initialize Latest
-  // Unrestricted: last cycle of last stage
-  const int Latest = -1;
+bool PostPipeliner::computeBackward() {
+  bool Changed = false;
+
+  auto AddOffspring = [&Changed](NodeInfo &Info, int E) {
+    if (Info.Offspring.insert(E).second) {
+      Changed = true;
+    }
+  };
+
+  // Traversing backwards will speed convergence a bit
+  for (int K = NInstr - 1; K >= 0; K--) {
+    SUnit &SU = DAG->SUnits[K];
+    auto &Me = Info[K];
+    const int Latest = Info[K].Latest;
+    for (auto &Dep : SU.Preds) {
+      if (Dep.getKind() != SDep::Data) {
+        continue;
+      }
+      int P = Dep.getSUnit()->NodeNum;
+      auto &Pred = Info[P];
+      AddOffspring(Pred, K);
+      for (auto Offs : Me.Offspring) {
+        AddOffspring(Pred, Offs);
+      }
+      int NewLatest = Latest - Dep.getSignedLatency();
+      if (NewLatest < Pred.Latest) {
+        Pred.Latest = NewLatest;
+        Changed = true;
+      }
+    }
+  }
+  return Changed;
+}
+
+bool PostPipeliner::computeLoopCarriedParameters() {
+
+  // Forward properties like Earliest and Ancestors.
+  computeForward();
+
+  // Backward properties like Latest and Offspring.
+  // Use a fixpoint loop, because plain reversed order may not be topological
+  // for predecessors
+  while (computeBackward()) {
+    /* EMPTY */;
+  }
+
+  // Adjust Earliest and Latest with resource requirements
+  for (int K = 0; K < NInstr; K++) {
+    auto &Me = Info[K];
+    SlotCounts ASlots;
+    for (int A : Me.Ancestors) {
+      auto &SU = DAG->SUnits[A];
+      ASlots += getSlotCounts(*SU.getInstr(), TII);
+    }
+    SlotCounts OSlots;
+    for (int O : Me.Offspring) {
+      auto &SU = DAG->SUnits[O];
+      OSlots += getSlotCounts(*SU.getInstr(), TII);
+    }
+    LLVM_DEBUG(dbgs() << "SU" << K << " : " << Info[K].Earliest << " - "
+                      << Info[K].Latest << " " << ASlots << " " << OSlots
+                      << "\n");
+    Me.Earliest = std::max(Me.Earliest, ASlots.max());
+    Me.Latest = std::min(Me.Latest, -OSlots.max());
```

Comment:
Not subjective whatsoever: minuses are confusing me :)

Reply:
It is nothing compared to the confusion that missing minus signs cause me Would you like me to make SlotCounts signed and supply a min() method?

Reply:
Checking again because I got confused: I think the computation is not symmetric.

As far as I understand, `Latest` belongs to `[-Length, -1]`, if our offsprings require e.g. 2 MOV slots, then (assuming a latency of 0 to those successors, and `Me` does not use the MOV slot), then the last feasible cycle is indeed `-2 = ASlots.max()`, because that is the second cycle counting from the bottom. 

For top-down, `Earliest` belongs to `[0, Length - 1]`. Using the same assumptions and 2 MOV slots booked by Ancestors, then `Me` should have an `Earlierst` of `1` which is the second cycle counting from the top. Not `ASlots.max()` which would be `2`. Just saying, but having positive "bottom-up cycles" like in the `MachineScheduler` would have effortlessly given symmetric results :trollface: 

What I would suggest is:
 - Include the slots of `Me` in `ASlots` and `OSlots`. Then we can get rid of the assumptions about latency and the slots booked by `Me`
 - Then we can compute:
   -  `EarliestCycle = MinRequiredCycles - 1 = ASlots.max() - 1`
   - `LatestCycle = LastCycle - (MinRequiredCycles - 1) = -1 - (OSlots.max() - 1) = -OSlots.max()`

Please double check me, my brain has been wired another way after looking into the MachineScheduler too much 😆

Reply:
Also, I think those computations assume no negative latencies. Otherwise, we cannot say much with SlotCounts because e.g. predecessors are not necessarily scheduled before `Me`. Maybe note that down, that we are aware of this, and still think that these metrics are good enough for the main cases?